video/out/wayland_common.c

```diff
@@ -1357,36 +1363,80 @@ static void format_table(void *data,
 {
     struct vo_wayland_state *wl = data;
 
+    if (wl->compositor_format_size) {
+        munmap(wl->compositor_format_map, wl->compositor_format_size);
+        wl->compositor_format_map = NULL;
+        wl->compositor_format_size = 0;
+    }
+
     void *map = mmap(NULL, size, PROT_READ, MAP_PRIVATE, fd, 0);
     close(fd);
 
     if (map != MAP_FAILED) {
-        wl->format_map = map;
-        wl->format_size = size;
+        wl->compositor_format_map = map;
+        wl->compositor_format_size = size;
     }
 }
 
 static void main_device(void *data,
                         struct zwp_linux_dmabuf_feedback_v1 *zwp_linux_dmabuf_feedback_v1,
                         struct wl_array *device)
 {
+    struct vo_wayland_state *wl = data;
+    wl->add_tranche = true;
+
 }
 
 static void tranche_done(void *data,
                          struct zwp_linux_dmabuf_feedback_v1 *zwp_linux_dmabuf_feedback_v1)
 {
+    struct vo_wayland_state *wl = data;
+    wl->add_tranche = false;
 }
 
 static void tranche_target_device(void *data,
                                   struct zwp_linux_dmabuf_feedback_v1 *zwp_linux_dmabuf_feedback_v1,
                                   struct wl_array *device)
 {
+    struct vo_wayland_state *wl = data;
+    // Only use the first tranche device we get.
+    if (wl->add_tranche) {
+        dev_t *id;
+        wl_array_for_each(id, device) {
+            memcpy(&wl->target_device_id, id, sizeof(dev_t));
+            break;
+        }
+        get_gpu_drm_formats(wl);
+    }
 }
 
 static void tranche_formats(void *data,
                             struct zwp_linux_dmabuf_feedback_v1 *zwp_linux_dmabuf_feedback_v1,
                             struct wl_array *indices)
 {
+    struct vo_wayland_state *wl = data;
+
+    // Only grab formats from the first tranche and ignore the rest.
+    if (!wl->add_tranche)
+        return;
+
+    assert(wl->compositor_format_map);
+    const compositor_format *formats = wl->compositor_format_map;
+
+    if (wl->compositor_formats)
+        talloc_free(wl->compositor_formats);
+    wl->compositor_formats = talloc_zero_array(wl, struct drm_format *, indices->size);
+
+    uint16_t *index;
+    wl_array_for_each(index, indices) {
+        struct drm_format *drm_format = talloc_zero(wl->compositor_formats, struct drm_format);
+        drm_format->format = formats[*index].format;
+        drm_format->modifier = formats[*index].modifier;
+        wl->compositor_formats[*index] = drm_format;
+        wl->num_compositor_formats++;
```

Comment:
Why not `MP_TARRAY_APPEND`?

Reply:
drm_format is small struct, I would just keep it in array storage instead of adding indirection.

Reply:
Why append when we already know the index beforehand? I know it doesn't really matter, but it's more work to have to loop to the new index everytime.

Reply:
> but it's more work to have to loop to the new index everytime.

It is not a list, it will do exactly the same as currently. Of course there is also `MP_TARRAY_INSERT_AT` to specify external index, but as I understand we add elements in order linearly.

Reply:
Just to be clear on what I mean, I think this would be an equivalent of current behavior 
``` diff
diff --git a/video/out/wayland_common.c b/video/out/wayland_common.c
index 6797864724..65aab736b3 100644
--- a/video/out/wayland_common.c
+++ b/video/out/wayland_common.c
@@ -1423,17 +1423,15 @@ static void tranche_formats(void *data,
     assert(wl->compositor_format_map);
     const compositor_format *formats = wl->compositor_format_map;
 
-    if (wl->compositor_formats)
-        talloc_free(wl->compositor_formats);
-    wl->compositor_formats = talloc_zero_array(wl, struct drm_format *, indices->size);
+    MP_RESIZE_ARRAY(wl, wl->compositor_formats, indices->size);
 
     uint16_t *index;
     wl_array_for_each(index, indices) {
-        struct drm_format *drm_format = talloc_zero(wl->compositor_formats, struct drm_format);
-        drm_format->format = formats[*index].format;
-        drm_format->modifier = formats[*index].modifier;
-        wl->compositor_formats[*index] = drm_format;
-        wl->num_compositor_formats++;
+        MP_TARRAY_APPEND(wl, wl->compositor_formats, wl->num_compositor_formats,
+                         (struct drm_format) {
+                            formats[*index].format,
+                            formats[*index].modifier,
+                        });
         MP_DBG(wl, "Compositor supports drm format: '%s(%016" PRIx64 ")'\n",
                mp_tag_str(drm_format->format), drm_format->modifier);
     }
diff --git a/video/out/wayland_common.h b/video/out/wayland_common.h
index 7db5571ec7..0ea4d8f805 100644
--- a/video/out/wayland_common.h
+++ b/video/out/wayland_common.h
@@ -113,7 +113,7 @@ struct vo_wayland_state {
     bool add_tranche;
     compositor_format *compositor_format_map;
     uint32_t compositor_format_size;
-    struct drm_format **compositor_formats;
+    struct drm_format *compositor_formats;
     int num_compositor_formats;
     uint32_t *gpu_formats;
     int num_gpu_formats;
diff --git a/video/out/wldmabuf/ra_wldmabuf.c b/video/out/wldmabuf/ra_wldmabuf.c
index b419a8662e..bd7f9abaad 100644
--- a/video/out/wldmabuf/ra_wldmabuf.c
+++ b/video/out/wldmabuf/ra_wldmabuf.c
@@ -32,7 +32,7 @@ bool ra_compatible_format(struct ra *ra, int imgfmt, uint32_t drm_format, uint64
 {
     struct priv *p = ra->priv;
     struct vo_wayland_state *wl = p->vo->wl;
-    struct drm_format **formats = wl->compositor_formats;
+    struct drm_format *formats = wl->compositor_formats;
 
     // First check if the gpu/drivers support the format.
     bool supported_gpu_format = false;
@@ -48,7 +48,7 @@ bool ra_compatible_format(struct ra *ra, int imgfmt, uint32_t drm_format, uint64
 
     // Now check if the compositor supports the format.
     for (int i = 0; i < wl->num_compositor_formats; ++i) {
-        if (drm_format == formats[i]->format && modifier == formats[i]->modifier)
+        if (drm_format == formats[i].format && modifier == formats[i].modifier)
             return true;
     }
 ```

Reply:
Oh yeah that works.